smart-contracts/contracts/mixins/MixinPurchase.sol

```diff
@@ -57,94 +57,106 @@ contract MixinPurchase is
-      if(expirationDuration == type(uint).max) {
-        newTimeStamp = type(uint).max;
+    require(_recipients.length == _referrers.length, 'INVALID_REFERRERS_LENGTH');
+    require(_recipients.length == _keyManagers.length, 'INVALID_KEY_MANAGERS_LENGTH');
+
+    uint totalPriceToPay;
+
+    for (uint256 i = 0; i < _recipients.length; i++) {
+      // check recipient address
+      address _recipient = _recipients[i];
+      require(_recipient != address(0), 'INVALID_ADDRESS');
+      
+      // Assign the key
+      Key storage toKey = keyByOwner[_recipient];
+      uint idTo = toKey.tokenId;
+      uint newTimeStamp;
+
+      if (idTo == 0) {
+        // Assign a new tokenId (if a new owner or previously transferred)
+        _assignNewTokenId(toKey);
+        // refresh the cached value
+        idTo = toKey.tokenId;
+        _recordOwner(_recipient, idTo);
+        // check for a non-expiring key
+        if (expirationDuration == type(uint).max) {
+          newTimeStamp = type(uint).max;
+        } else {
+          newTimeStamp = block.timestamp + expirationDuration;
+        }
+        toKey.expirationTimestamp = newTimeStamp;
+
+        // set key manager
+        _setKeyManagerOf(idTo, _keyManagers[i]);
+
+        // trigger event
+        emit Transfer(
+          address(0), // This is a creation.
+          _recipient,
+          idTo
+        );
+      } else if (toKey.expirationTimestamp > block.timestamp) {
+        // prevent re-purchase of a valid non-expiring key
+        require(toKey.expirationTimestamp != type(uint).max, 'A valid non-expiring key can not be purchased twice');
+
+        // This is an existing owner trying to extend their key
+        newTimeStamp = toKey.expirationTimestamp + expirationDuration;
+        toKey.expirationTimestamp = newTimeStamp;
+
+        emit RenewKeyPurchase(_recipient, newTimeStamp);
       } else {
-        newTimeStamp = block.timestamp + expirationDuration;
-      }
-      toKey.expirationTimestamp = newTimeStamp;
+        // This is an existing owner trying to renew their expired or cancelled key
+        if(expirationDuration == type(uint).max) {
+          newTimeStamp = type(uint).max;
+        } else {
+          newTimeStamp = block.timestamp + expirationDuration;
+        }
+        toKey.expirationTimestamp = newTimeStamp;
 
-      _setKeyManagerOf(idTo, _keyManager);
+        _setKeyManagerOf(idTo, _keyManagers[i]);
 
-      emit RenewKeyPurchase(_recipient, newTimeStamp);
-    }
+        emit RenewKeyPurchase(_recipient, newTimeStamp);
+      }
 
-    
-    uint inMemoryKeyPrice = _purchasePriceFor(_recipient, _referrer, _data);
+      // price      
+      uint inMemoryKeyPrice = _purchasePriceFor(_recipient, _referrers[i], _data);
+      totalPriceToPay = totalPriceToPay + inMemoryKeyPrice;
+
+      // call Unlock contract to record GNP
+      try unlockProtocol.recordKeyPurchase(inMemoryKeyPrice, _referrers[i]) 
```

Comment:
is that were we should check the gas limit too?

Reply:
the gas limit could be checked against a max number of purchases that can be made at once, so we check as `require(_recipients.length == 20, ''20_PURCHASES_MAX");`

Reply:
No no i meant the attack vector identified during the audit :) where unlock is compromised but could not break the locks themselves.